src/include/common/object_pool.h

```diff
@@ -38,23 +68,74 @@ class ObjectPool {
    */
   ~ObjectPool() {
     T *result = nullptr;
-    while (reuse_queue_.Dequeue(&result)) alloc_.Delete(result);
+    while (!reuse_queue_.empty()) {
+      result = reuse_queue_.front();
+      alloc_.Delete(result);
+      reuse_queue_.pop();
+    }
   }
 
   /**
    * Returns a piece of memory to hold an object of T.
-   *
+   * @throw NoMoreObjectException if the object pool fails to fetch memory.
    * @return pointer to memory that can hold T
    */
   T *Get() {
+    SpinLatch::ScopedSpinLatch guard(&latch_);
     T *result = nullptr;
-    if (!reuse_queue_.Dequeue(&result))
-      result = alloc_.New();
-    else
+    if (reuse_queue_.empty()) {
+      if (current_size_ < size_limit_) result = alloc_.New();
+      if (result == nullptr) {
+        // out of memory
+        throw NoMoreObjectException(size_limit_);
+      }
+      current_size_++;
+    } else {
+      result = reuse_queue_.front();
+      reuse_queue_.pop();
       alloc_.Reuse(result);
+    }
+    TERRIER_ASSERT(current_size_ <= size_limit_, "object pool size exceed it's size limit");
     return result;
   }
 
+  /**
+   * Set the object pool's size limit.
+   *
+   * The operation fails if the object pool has already allocated more objects
+   * than the size limit.
+   *
+   * @param new_size the new object pool size
+   * @return true if new_size is successfully set and false the operation fails
+   */
+  bool SetSizeLimit(uint64_t new_size) {
+    SpinLatch::ScopedSpinLatch guard(&latch_);
+    if (new_size >= current_size_) {
+      // current_size_ might increase and become > new_size if we don't use lock
+      size_limit_ = new_size;
+      TERRIER_ASSERT(current_size_ <= size_limit_, "object pool size exceed it's size limit");
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Set the reuse limit to a new value.
```

Comment:
also document what the function will do. (It doesn't have to be an exact description of behavior, but some level of contract is needed: does it always happen immediately? Or is it not guaranteed? Is it guaranteed to be done by some point?)